File: internal/pluginfixture/service.go
package pluginfixture

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/json"
	"errors"
	"io"
	"runtime"
	"slices"
	"sync"

	commonv1 "github.com/spice-framework/spice-agent/common/v1"
	pluginv1 "github.com/spice-framework/spice-agent/plugin/v1"
	"github.com/spice-framework/spice-agent/tool"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/encoding/protowire"
)

const (
	maximumFixtureMessageBytes = pluginv1.InitializeBootstrapMaximumBytes
	echoToolName               = "conformance.echo"
	failToolName               = "conformance.fail"
	waitToolName               = "conformance.wait"
)

var fixtureUnknownField = func() []byte {
	value := protowire.AppendTag(nil, 127, protowire.BytesType)
	return protowire.AppendBytes(value, []byte("fixture-compatible"))
}()

type service struct {
	pluginv1.UnimplementedPluginServiceServer

	secret   []byte
	limits   *pluginv1.Limits
	manifest *pluginv1.Manifest
	shutdown func()

	mu          sync.Mutex
	initialized bool
	draining    bool
	closed      bool
	sessionID   []byte
	active      uint32
	zero        chan struct{}
}

// NewService constructs the independent Go conformance fixture service. It is
// internal test infrastructure, not the production runtime-plugin host.
func NewService(secret []byte, shutdown func()) (pluginv1.PluginServiceServer, error) {
	if len(secret) != pluginv1.HandshakeSecretBytes || shutdown == nil {
		return nil, errors.New("fixture secret and shutdown callback are required")
	}
	manifest, err := fixtureManifest()
	if err != nil {
		return nil, err
	}
	zero := make(chan struct{})
	close(zero)
	return &service{
		secret:   slices.Clone(secret),
		limits:   fixtureLimits(),
		manifest: manifest,
		shutdown: shutdown,
		zero:     zero,
	}, nil
}

func (current *service) Initialize(
	_ context.Context,
	request *pluginv1.InitializeRequest,
) (*pluginv1.InitializeResponse, error) {
	if err := pluginv1.ValidateInitializeRequest(request); err != nil {
		return current.failedInitialize(request, commonv1.ErrorCode_ERROR_CODE_INVALID_ARGUMENT,
			"plugin initialization request is invalid")
	}
	current.mu.Lock()
	defer current.mu.Unlock()
	if current.initialized {
		return current.failedInitialize(request, commonv1.ErrorCode_ERROR_CODE_CONFLICT,
			"plugin session is already initialized")
	}
	protocol, protocolStatus := commonv1.NegotiateProtocol(request.GetProtocol(), pluginv1.SupportedProtocolRange())
	if protocolStatus.GetCode() != commonv1.ErrorCode_ERROR_CODE_OK {
		return current.failedInitialize(request, protocolStatus.GetCode(), protocolStatus.GetMessage())
	}
	capabilities, capabilityStatus := commonv1.NegotiateCapabilities(
		request.GetSupportedCapabilities(),
		request.GetRequiredCapabilities(),
		&commonv1.CapabilitySet{Names: []string{pluginv1.CapabilityRuntimeToolsV1}},
	)
	if capabilityStatus.GetCode() != commonv1.ErrorCode_ERROR_CODE_OK {
		return current.failedInitialize(request, capabilityStatus.GetCode(), capabilityStatus.GetMessage())
	}
	limits, err := pluginv1.NegotiateLimits(request.GetRequestedLimits(), current.limits)
	if err != nil {
		return current.failedInitialize(request, commonv1.ErrorCode_ERROR_CODE_INVALID_ARGUMENT,
			"plugin limit negotiation failed")
	}
	if _, err = pluginv1.DecodeManifest(current.manifest, limits); err != nil {
		return current.failedInitialize(request, commonv1.ErrorCode_ERROR_CODE_INVALID_ARGUMENT,
			"plugin manifest exceeds negotiated limits")
	}
	sessionID := make([]byte, pluginv1.SessionIDBytes)
	if _, err = io.ReadFull(rand.Reader, sessionID); err != nil {
		return nil, status.Error(codes.Internal, "plugin session identity generation failed")
	}
	response := &pluginv1.InitializeResponse{
		Status:       commonv1.OKStatus(),
		Protocol:     protocol,
		Plugin:       fixtureBuild(),
		Capabilities: capabilities,
		Limits:       limits,
		Manifest:     current.manifest,
		LaunchId:     slices.Clone(request.GetLaunchId()),
		SessionId:    sessionID,
		HandshakeChallenge: slices.Clone(
			request.GetHandshakeChallenge(),
		),
	}
	signed, err := pluginv1.SignInitializeResponse(request, response, current.secret)
	if err != nil {
		return nil, status.Error(codes.Internal, "plugin handshake signing failed")
	}
	current.initialized = true
	current.limits = cloneLimits(limits)
	current.sessionID = slices.Clone(sessionID)
	return signed, nil
}

func (current *service) failedInitialize(
	request *pluginv1.InitializeRequest,
	code commonv1.ErrorCode,
	message string,
) (*pluginv1.InitializeResponse, error) {
	if request == nil || len(request.GetLaunchId()) != pluginv1.LaunchIDBytes ||
		len(request.GetHandshakeChallenge()) != pluginv1.HandshakeChallengeBytes {
		return nil, status.Error(codes.InvalidArgument, "plugin initialization request is invalid")
	}
	response := &pluginv1.InitializeResponse{
		Status: &commonv1.Status{Code: code, Message: message},
		Plugin: fixtureBuild(),
		LaunchId: slices.Clone(
			request.GetLaunchId(),
		),
		HandshakeChallenge: slices.Clone(request.GetHandshakeChallenge()),
	}
	signed, err := pluginv1.SignInitializeResponse(request, response, current.secret)
	if err != nil {
		return nil, status.Error(codes.Internal, "plugin handshake signing failed")
	}
	return signed, nil
}

func (current *service) Execute(
	request *pluginv1.ExecuteRequest,
	stream pluginv1.PluginService_ExecuteServer,
) error {
	if request == nil {
		return status.Error(codes.InvalidArgument, "plugin call is invalid")
	}
	current.mu.Lock()
	if !current.initialized || current.closed {
		current.mu.Unlock()
		return status.Error(codes.FailedPrecondition, "plugin session is unavailable")
	}
	if current.draining {
		current.mu.Unlock()
		return status.Error(codes.Unavailable, "plugin is draining")
	}
	if uint64(len(request.GetArgumentsJson())) > current.limits.GetMaxCallArgumentBytes() {
		current.mu.Unlock()
		return status.Error(codes.ResourceExhausted, "plugin call exceeds the negotiated limit")
	}
	call, err := pluginv1.DecodeExecuteRequest(request, current.sessionID, current.limits)
	if err != nil {
		current.mu.Unlock()
		return status.Error(codes.InvalidArgument, "plugin call is invalid")
	}
	if !knownTool(call.Name()) {
		current.mu.Unlock()
		return status.Error(codes.NotFound, "plugin tool is unavailable")
	}
	if current.active >= current.limits.GetMaxConcurrentCalls() {
		current.mu.Unlock()
		return status.Error(codes.ResourceExhausted, "plugin concurrent-call limit is exhausted")
	}
	current.admitLocked()
	current.mu.Unlock()
	defer current.release()

	switch call.Name() {
	case echoToolName:
		return executeEcho(call, stream)
	case failToolName:
		return executeFailure(call, stream)
	case waitToolName:
		return executeWait(call, stream)
	default:
		return status.Error(codes.Internal, "plugin tool dispatch is inconsistent")
	}
}

func (current *service) Drain(
	ctx context.Context,
	request *pluginv1.DrainRequest,
) (*pluginv1.DrainResponse, error) {
	current.mu.Lock()
	if !current.initialized || current.closed {
		current.mu.Unlock()
		return nil, status.Error(codes.FailedPrecondition, "plugin session is unavailable")
	}
	if err := pluginv1.ValidateDrainRequest(request, current.sessionID, current.limits); err != nil {
		current.mu.Unlock()
		return nil, status.Error(codes.InvalidArgument, "plugin drain request is invalid")
	}
	current.draining = true
	zero := current.zero
	current.mu.Unlock()
	select {
	case <-ctx.Done():
		return nil, status.FromContextError(ctx.Err()).Err()
	case <-zero:
		return &pluginv1.DrainResponse{Status: commonv1.OKStatus()}, nil
	}
}

func (current *service) Shutdown(
	_ context.Context,
	request *pluginv1.ShutdownRequest,
) (*pluginv1.ShutdownResponse, error) {
	current.mu.Lock()
	if !current.initialized || current.closed {
		current.mu.Unlock()
		return nil, status.Error(codes.FailedPrecondition, "plugin session is unavailable")
	}
	if err := pluginv1.ValidateShutdownRequest(request, current.sessionID, current.limits); err != nil {
		current.mu.Unlock()
		return nil, status.Error(codes.InvalidArgument, "plugin shutdown request is invalid")
	}
	if !current.draining || current.active != 0 {
		current.mu.Unlock()
		return nil, status.Error(codes.FailedPrecondition, "plugin must drain before shutdown")
	}
	current.closed = true
	shutdown := current.shutdown
	current.mu.Unlock()
	go shutdown()
	return &pluginv1.ShutdownResponse{Status: commonv1.OKStatus()}, nil
}

func (current *service) admitLocked() {
	if current.active == 0 {
		current.zero = make(chan struct{})
	}
	current.active++
}

func (current *service) release() {
	current.mu.Lock()
	defer current.mu.Unlock()
	current.active--
	if current.active == 0 {
		close(current.zero)
	}
}

func executeEcho(call tool.Call, stream pluginv1.PluginService_ExecuteServer) error {
	var input struct {
		Value string `json:"value"`
	}
	if err := json.Unmarshal(call.Arguments(), &input); err != nil || input.Value == "" {
		return status.Error(codes.InvalidArgument, "plugin echo arguments are invalid")
	}
	progress := &pluginv1.ExecuteResponse{
		CallId: string(call.ID()), Sequence: 1,
		Frame: &pluginv1.ExecuteResponse_Progress{Progress: &pluginv1.Progress{Message: "echo accepted"}},
	}
	progress.ProtoReflect().SetUnknown(slices.Clone(fixtureUnknownField))
	if err := stream.Send(progress); err != nil {
		return err
	}
	content, err := json.Marshal(input)
	if err != nil {
		return status.Error(codes.Internal, "plugin echo encoding failed")
	}
	return stream.Send(&pluginv1.ExecuteResponse{
		CallId: string(call.ID()), Sequence: 2,
		Frame: &pluginv1.ExecuteResponse_Result{Result: &pluginv1.Result{ContentJson: content}},
	})
}

func executeFailure(call tool.Call, stream pluginv1.PluginService_ExecuteServer) error {
	if !bytes.Equal(call.Arguments(), []byte(`{}`)) {
		return status.Error(codes.InvalidArgument, "plugin failure arguments are invalid")
	}
	return stream.Send(&pluginv1.ExecuteResponse{
		CallId: string(call.ID()), Sequence: 1,
		Frame: &pluginv1.ExecuteResponse_Failure{Failure: &pluginv1.ExecutionFailure{
			State:       pluginv1.ExecutionState_EXECUTION_STATE_DEFINITIVE,
			Retry:       pluginv1.RetryDisposition_RETRY_DISPOSITION_NEVER,
			SafeMessage: "fixture failure",
		}},
	})
}

func executeWait(call tool.Call, stream pluginv1.PluginService_ExecuteServer) error {
	if !bytes.Equal(call.Arguments(), []byte(`{}`)) {
		return status.Error(codes.InvalidArgument, "plugin wait arguments are invalid")
	}
	if err := stream.Send(&pluginv1.ExecuteResponse{
		CallId: string(call.ID()), Sequence: 1,
		Frame: &pluginv1.ExecuteResponse_Progress{Progress: &pluginv1.Progress{Message: "waiting"}},
	}); err != nil {
		return err
	}
	<-stream.Context().Done()
	return status.FromContextError(stream.Context().Err()).Err()
}

func fixtureManifest() (*pluginv1.Manifest, error) {
	definitions := make([]tool.Definition, 0, 3)
	for _, item := range []struct {
		name        string
		description string
		schema      string
	}{
		{echoToolName, "Echo one string value.", `{"type":"object","properties":{"value":{"type":"string"}},"required":["value"]}`},
		{failToolName, "Return one definitive fixture failure.", `{"type":"object"}`},
		{waitToolName, "Wait until the call is cancelled.", `{"type":"object"}`},
	} {
		definition, err := tool.NewDefinition(
			item.name,
			item.description,
			json.RawMessage(item.schema),
			tool.EffectReadOnly,
			tool.ReplaySafe,
		)
		if err != nil {
			return nil, err
		}
		definitions = append(definitions, definition)
	}
	catalog, err := pluginv1.NewCatalog("spice-agent-go-conformance", "v1", definitions, fixtureLimits())
	if err != nil {
		return nil, err
	}
	return catalog.Manifest()
}

func fixtureLimits() *pluginv1.Limits {
	return &pluginv1.Limits{
		MaxMessageBytes:      maximumFixtureMessageBytes,
		MaxTools:             16,
		MaxSchemaBytes:       64 << 10,
		MaxCallArgumentBytes: 64 << 10,
		MaxResultBytes:       maximumFixtureMessageBytes,
		MaxProgressBytes:     tool.MaximumProgressBytes,
		MaxConcurrentCalls:   8,
	}
}

func fixtureBuild() *pluginv1.BuildIdentity {
	return &pluginv1.BuildIdentity{
		Component: "spice-agent-go-conformance",
		Version:   "v1",
		Commit:    "fixture",
		Runtime:   runtime.Version(),
	}
}

func knownTool(name string) bool {
	return name == echoToolName || name == failToolName || name == waitToolName
}

func cloneLimits(value *pluginv1.Limits) *pluginv1.Limits {
	return &pluginv1.Limits{
		MaxMessageBytes:      value.GetMaxMessageBytes(),
		MaxTools:             value.GetMaxTools(),
		MaxSchemaBytes:       value.GetMaxSchemaBytes(),
		MaxCallArgumentBytes: value.GetMaxCallArgumentBytes(),
		MaxResultBytes:       value.GetMaxResultBytes(),
		MaxProgressBytes:     value.GetMaxProgressBytes(),
		MaxConcurrentCalls:   value.GetMaxConcurrentCalls(),
	}
}

File: testdata/runtimeplugin/python/src/spice_agent_python_fixture/service.py
"""Async gRPC implementation of the frozen runtime-tool conformance profile."""

from __future__ import annotations

import asyncio
import json
import secrets
import sys
from collections.abc import Callable

import grpc
from spice.agent.common.v1 import common_pb2
from spice.agent.plugin.v1 import plugin_pb2, plugin_pb2_grpc

from . import protocol

ECHO_TOOL = "conformance.echo"
FAIL_TOOL = "conformance.fail"
WAIT_TOOL = "conformance.wait"


class PluginService(plugin_pb2_grpc.PluginServiceServicer):
    """One connection-owned plugin generation used by the public conformance kit."""

    def __init__(self, secret: bytearray, shutdown: Callable[[], None]) -> None:
        if len(secret) != protocol.HANDSHAKE_SECRET_BYTES:
            raise ValueError("fixture secret has an invalid size")
        self._secret = secret
        self._shutdown = shutdown
        self._limits = protocol.fixture_limits()
        self._manifest = _fixture_manifest()
        self._lock = asyncio.Lock()
        self._zero_active = asyncio.Event()
        self._zero_active.set()
        self._initialized = False
        self._draining = False
        self._closed = False
        self._session_id = b""
        self._active = 0

    async def Initialize(self, request, context):  # noqa: N802 - generated RPC API
        try:
            protocol.validate_initialize(request)
        except protocol.RequestError:
            return self._failed_initialize(
                request,
                common_pb2.ERROR_CODE_INVALID_ARGUMENT,
                "plugin initialization request is invalid",
                context,
            )

        async with self._lock:
            if self._initialized:
                return self._failed_initialize(
                    request,
                    common_pb2.ERROR_CODE_CONFLICT,
                    "plugin session is already initialized",
                    context,
                )
            selected_protocol = protocol.select_protocol(request.protocol)
            if selected_protocol is None:
                return self._failed_initialize(
                    request,
                    common_pb2.ERROR_CODE_INCOMPATIBLE_VERSION,
                    "client and plugin protocol ranges do not overlap",
                    context,
                )
            selected_capabilities = protocol.select_capabilities(request)
            if selected_capabilities is None:
                return self._failed_initialize(
                    request,
                    common_pb2.ERROR_CODE_MISSING_CAPABILITY,
                    "required client capabilities are unavailable",
                    context,
                )
            try:
                selected_limits = protocol.negotiate_limits(request.requested_limits)
                protocol.validate_manifest(self._manifest, selected_limits)
            except protocol.RequestError:
                return self._failed_initialize(
                    request,
                    common_pb2.ERROR_CODE_INVALID_ARGUMENT,
                    "plugin limit or manifest negotiation failed",
                    context,
                )

            session_id = secrets.token_bytes(protocol.SESSION_ID_BYTES)
            response = plugin_pb2.InitializeResponse(
                status=protocol.ok_status(),
                protocol=selected_protocol,
                plugin=_fixture_build(),
                capabilities=common_pb2.CapabilitySet(names=selected_capabilities),
                limits=selected_limits,
                manifest=self._manifest,
                launch_id=request.launch_id,
                session_id=session_id,
                handshake_challenge=request.handshake_challenge,
            )
            signed = protocol.sign_initialize(request, response, self._secret)
            self._initialized = True
            self._limits = protocol.clone_limits(selected_limits)
            self._session_id = session_id
            return signed

    async def Execute(self, request, context):  # noqa: N802 - generated RPC API
        async with self._lock:
            if not self._initialized or self._closed:
                await context.abort(grpc.StatusCode.FAILED_PRECONDITION, "plugin session is unavailable")
            if self._draining:
                await context.abort(grpc.StatusCode.UNAVAILABLE, "plugin is draining")
            try:
                protocol.validate_execute(request, self._session_id, self._limits)
            except OverflowError:
                await context.abort(
                    grpc.StatusCode.RESOURCE_EXHAUSTED,
                    "plugin call exceeds the negotiated limit",
                )
            except protocol.RequestError:
                await context.abort(grpc.StatusCode.INVALID_ARGUMENT, "plugin call is invalid")
            if request.tool_name not in {ECHO_TOOL, FAIL_TOOL, WAIT_TOOL}:
                await context.abort(grpc.StatusCode.NOT_FOUND, "plugin tool is unavailable")
            if self._active >= self._limits.max_concurrent_calls:
                await context.abort(
                    grpc.StatusCode.RESOURCE_EXHAUSTED,
                    "plugin concurrent-call limit is exhausted",
                )
            self._active += 1
            self._zero_active.clear()

        try:
            if request.tool_name == ECHO_TOOL:
                async for response in self._execute_echo(request, context):
                    yield response
            elif request.tool_name == FAIL_TOOL:
                yield await self._execute_failure(request, context)
            else:
                async for response in self._execute_wait(request, context):
                    yield response
        finally:
            async with self._lock:
                self._active -= 1
                if self._active == 0:
                    self._zero_active.set()

    async def Drain(self, request, context):  # noqa: N802 - generated RPC API
        async with self._lock:
            await self._require_session(request.session_id, context)
            self._draining = True
            zero_active = self._zero_active
        await zero_active.wait()
        return plugin_pb2.DrainResponse(status=protocol.ok_status(), active_calls=0)

    async def Shutdown(self, request, context):  # noqa: N802 - generated RPC API
        async with self._lock:
            await self._require_session(request.session_id, context)
            if not self._draining or self._active != 0:
                await context.abort(
                    grpc.StatusCode.FAILED_PRECONDITION,
                    "plugin must drain before shutdown",
                )
            self._closed = True
            for index in range(len(self._secret)):
                self._secret[index] = 0
        asyncio.get_running_loop().call_later(0.05, self._shutdown)
        return plugin_pb2.ShutdownResponse(status=protocol.ok_status())

    def _failed_initialize(self, request, code: int, message: str, context):
        if (
            len(request.launch_id) != protocol.LAUNCH_ID_BYTES
            or len(request.handshake_challenge) != protocol.HANDSHAKE_CHALLENGE_BYTES
        ):
            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
            context.set_details("plugin initialization request is invalid")
            return plugin_pb2.InitializeResponse()
        response = plugin_pb2.InitializeResponse(
            status=protocol.error_status(code, message),
            plugin=_fixture_build(),
            launch_id=request.launch_id,
            handshake_challenge=request.handshake_challenge,
        )
        return protocol.sign_initialize(request, response, self._secret)

    async def _require_session(self, observed: bytes, context) -> None:
        if not self._initialized or self._closed:
            await context.abort(
                grpc.StatusCode.FAILED_PRECONDITION, "plugin session is unavailable"
            )
        try:
            protocol.validate_session(observed, self._session_id)
        except protocol.RequestError:
            await context.abort(grpc.StatusCode.INVALID_ARGUMENT, "plugin session is invalid")

    async def _execute_echo(self, request, context):
        try:
            value = json.loads(request.arguments_json)
        except (UnicodeDecodeError, json.JSONDecodeError):
            await context.abort(grpc.StatusCode.INVALID_ARGUMENT, "plugin echo arguments are invalid")
        if not isinstance(value, dict) or not isinstance(value.get("value"), str) or not value["value"]:
            await context.abort(grpc.StatusCode.INVALID_ARGUMENT, "plugin echo arguments are invalid")
        progress = plugin_pb2.ExecuteResponse(
            call_id=request.call_id,
            sequence=1,
            progress=plugin_pb2.Progress(message="echo accepted"),
        )
        protocol.add_unknown_bytes(progress, 127, b"fixture-compatible")
        yield progress
        content = json.dumps(
            {"value": value["value"]}, separators=(",", ":"), ensure_ascii=False
        ).encode("utf-8")
        yield plugin_pb2.ExecuteResponse(
            call_id=request.call_id,
            sequence=2,
            result=plugin_pb2.Result(content_json=content),
        )

    async def _execute_failure(self, request, context):
        if request.arguments_json != b"{}":
            await context.abort(
                grpc.StatusCode.INVALID_ARGUMENT, "plugin failure arguments are invalid"
            )
        return plugin_pb2.ExecuteResponse(
            call_id=request.call_id,
            sequence=1,
            failure=plugin_pb2.ExecutionFailure(
                state=plugin_pb2.EXECUTION_STATE_DEFINITIVE,
                retry=plugin_pb2.RETRY_DISPOSITION_NEVER,
                safe_message="fixture failure",
            ),
        )

    async def _execute_wait(self, request, context):
        if request.arguments_json != b"{}":
            await context.abort(grpc.StatusCode.INVALID_ARGUMENT, "plugin wait arguments are invalid")
        yield plugin_pb2.ExecuteResponse(
            call_id=request.call_id,
            sequence=1,
            progress=plugin_pb2.Progress(message="waiting"),
        )
        await asyncio.Future()


def _fixture_manifest() -> plugin_pb2.Manifest:
    definitions = []
    for name, description, schema in (
        (
            ECHO_TOOL,
            "Echo one string value.",
            '{"type":"object","properties":{"value":{"type":"string"}},"required":["value"]}',
        ),
        (FAIL_TOOL, "Return one definitive fixture failure.", '{"type":"object"}'),
        (WAIT_TOOL, "Wait until the call is cancelled.", '{"type":"object"}'),
    ):
        definitions.append(
            plugin_pb2.ToolDefinition(
                name=name,
                description=description,
                input_schema_json=schema.encode("utf-8"),
                effect=plugin_pb2.TOOL_EFFECT_READ_ONLY,
                replay_safety=plugin_pb2.REPLAY_SAFETY_SAFE,
                capabilities=common_pb2.CapabilitySet(),
            )
        )
    return plugin_pb2.Manifest(
        name="spice-agent-python-conformance",
        version="v1",
        tools=definitions,
    )


def _fixture_build() -> plugin_pb2.BuildIdentity:
    return plugin_pb2.BuildIdentity(
        component="spice-agent-python-conformance",
        version="v1",
        commit="fixture",
        runtime=f"python{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
    )
